bigquery/link-scraped-vacancies-to-schools-and-vacancies.sql

```diff
@@ -102,31 +108,91 @@ WITH
     #exclude matches with schools which were closed or had not yet been created in GIAS on the date when the vacancy was published (note - will exclude a match if a school published a vacancy during academisation - i.e. after the new academy had been created in GIAS but before the old school had officially closed - because in this case we can't work out whether recruitment was for the old or the new school
       (school.data_establishmentstatus_name != "Closed"
         OR school.data_closedate > scraped_vacancy.publish_on)
-      AND CAST(school.created_at AS DATE) <= scraped_vacancy.publish_on )
+      AND CAST(school.created_at AS DATE) <= scraped_vacancy.publish_on ),
+    vacancies_joined_to_schools AS (
+    SELECT
+      scraped_vacancy.*,
+      vacancy_matches.school_urn,
+      school.name AS school_name
+    FROM
+      scraped_vacancy
+    LEFT JOIN
+      vacancy_matches
+    ON
+      scraped_vacancy.scraped_url=vacancy_matches.scraped_url
+    LEFT JOIN
+      `teacher-vacancy-service.production_dataset.feb20_school` AS school
+    ON
+      school_urn=school.urn
+    WHERE
+      scraped_vacancy.scraped_url IN ( # only include matches with only 1 possible school
+      SELECT
+        scraped_url AS scraped_url,
+      FROM
+        vacancy_matches
+      GROUP BY
+        scraped_url
+      HAVING
+        COUNT(*)=1 )
+      OR vacancy_matches.school_urn IS NULL
+    ORDER BY
+      scraped_url ASC ),
+    vacancy_to_vacancy_matches AS (
+    SELECT
+      vacancies_joined_to_schools.scraped_url,
+      TV_vacancy.id AS vacancy_id
+    FROM
+      vacancies_joined_to_schools
+    LEFT JOIN
+      `teacher-vacancy-service.production_dataset.feb20_school` AS school
+    ON
+      vacancies_joined_to_schools.school_urn=school.urn
+    LEFT JOIN
+      `teacher-vacancy-service.production_dataset.feb20_vacancy` AS TV_vacancy
+    ON
+      school.id=TV_vacancy.school_id #only match vacancies to TV vacancies from the matched school in our database
+      AND ( (
+        SELECT
+          SAFE_DIVIDE(COUNTIF(scraped_words IS NULL
+          OR TV_words IS NULL),COUNTIF(scraped_words IS NOT NULL
+          AND TV_words IS NOT NULL)) #number of words that are in one title but not in the other as a proportion of the number of words that are in both titles
+        FROM (
+          SELECT
+            *
+          FROM
+            UNNEST(SPLIT(LOWER(REGEXP_REPLACE(TV_vacancy.job_title,r'[^a-zA-Z0-9]', '')),' ')) AS words
+          WHERE words NOT IN ('at','of','to','and','or','for','with','but','may','be','as','x','from','including','per','in') #don't count these words when working out the proportion
```

Comment:
What happens if we also exclude 'teacher'? My thinking is that 'teacher' is a common but generally meaningless word in this context.

Reply:
For avoidance of duplication, it would be a nice-to-have for the list of stop-words to have a single source of truth. Not sure how easy that is in SQL.

Reply:
Thanks @david-mears-dfe - 'teacher' is actually quite important hear, because we need to avoid matching teacher roles to teaching assistant roles or headteacher roles, for example. Many primary jobs are in fact just listed as 'Teacher'.

I'll look into the list of stop words. Something might be possible, although it might be clunky!

Reply:
@david-mears-dfe - sorted the list of stop words.